src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/CpsDiagnosticItemSource.cs

```diff
@@ -31,9 +31,29 @@ public CpsDiagnosticItemSource(Workspace workspace, string projectPath, ProjectI
             _analyzerReference = TryGetAnalyzerReference(Workspace.CurrentSolution);
             if (_analyzerReference == null)
             {
-                // The workspace doesn't know about the project and/or the analyzer yet.
-                // Hook up an event handler so we can update when it does.
-                Workspace.WorkspaceChanged += OnWorkspaceChangedLookForAnalyzer;
+                // The ProjectId that was given to us was found by enumerating the list of projects in the solution, thus the project must have already
+                // been added to the workspace at some point. As long as the project is still there, we're going to assume the reason we don't have the reference
+                // yet is because while we have a project, we don't have all the references added yet. We'll wait until we see the reference and then connect to it.
+                if (workspace.CurrentSolution.ContainsProject(projectId))
```

Comment:
Theoretically, would it be possible the project change event before fired between calling `TryGetAnalyzerReference` above and the subscription in the `if` below, therefore we missed the opportunity to update `_analyzerReference`? 
If so, does it matter?

Reply:
Ah, absolutely yes. Good catch! We've had some sporadic reports of this UI not working, so I'll patch this to make sure it's correct. I don't like this code generally (and we're going to have to revisit it) but I'd like there to at least not be known issues with it which can make tracking down new issues harder.

Reply:
@genlu Updated!